File: sdk/unity/Assets/Falken/Scripts/Entity.cs
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace Falken
{

    /// <summary>
    /// <c>EntityBase</c> Observations.
    /// </summary>
    public class EntityBase : Falken.AttributeContainer
    {
        /// <summary>
        /// <c>EntityNotBoundException</c> thrown when access to an
        /// unbound entity property is forbidden.
        /// </summary>
        [Serializable]
        public class EntityNotBoundException : Exception
        {
            internal EntityNotBoundException()
            {
            }

            internal EntityNotBoundException(string message)
              : base(message) { }

            internal EntityNotBoundException(string message, Exception inner)
              : base(message, inner) { }
        }

        // Do not warn about this field being unused as we are explicitly holding a reference to an
        // unmanaged object.
#pragma warning disable 0414
        // Internal falken's entity that this instance
        // is linked to.
        private FalkenInternal.falken.EntityBase _entity = null;
#pragma warning restore 0414

        // Name used for dynamic entities. Will set to null
        // once this entity is bound.
        private string _name = null;

        /// <summary>
        /// Position in world space.
        /// </summary>
        [FalkenInherited]
        public Falken.PositionVector position = new Falken.PositionVector();

        /// <summary>
        /// Rotation in world space.
        /// </summary>
        [FalkenInherited]
        public Falken.RotationQuaternion rotation = new Falken.RotationQuaternion();

        /// <summary>
        /// Check if the entity is bound to a Falken's entity.
        /// </summary>
        internal new bool Bound
        {
            get
            {
                return _entity != null;
            }
        }

        /// <summary>
        /// Create a empty Falken.Entity that is bound to a field.
        /// </summary>
        public EntityBase()
        {
        }

        /// <summary>
        /// Create a dynamic Falken.Entity.
        /// <exception>
        /// ArgumentNullException thrown if name is null.
        /// </exception>
        /// </summary>
        public EntityBase(string name)
        {
            if (name == null)
            {
                throw new ArgumentNullException(name);
            }
            _name = name;
        }

        /// <summary>
        /// Get the name of this entity (which will be equal to the field name).
        /// </summary>
        public string Name
        {
            get
            {
                if (Bound)
                {
                    return _entity.name();
                }
                throw new EntityNotBoundException(
                  "Can't retrieve the name if the entity is not bound.");
            }
        }

        /// <summary>
        /// Establish a connection for all the C# attributes defined
        /// in this instance.
        /// <exception> AlreadyBoundException thrown when trying to
        /// bind the entity when it was bound already. </exception>
        /// </summary>
        internal void BindEntity(FieldInfo field,
          FalkenInternal.falken.EntityContainer container)
        {
            if (Bound)
            {
                throw new AlreadyBoundException(
                  "Can't bind entity more than once.");
            }
            FalkenInternal.falken.EntityBase entity = null;
            if (field != null)
            {
                if (System.Attribute.GetCustomAttribute(field,
                    typeof(FalkenInheritedAttribute)) == null)
                {
                    entity = new FalkenInternal.falken.EntityBase(container,
                      field.Name);
                }
                else
                {
                    entity = container.entity(field.Name);
                    if (entity == null)
                    {
                        throw new InheritedAttributeNotFoundException(
                          $"Inherited entity field '{field.Name}' was not found " +
                          $"in the entity container.");
                    }
                }
            }
            else
            {
                entity = new FalkenInternal.falken.EntityBase(container, _name);
                _name = null;
            }
            base.BindAttributes(entity);
            _entity = entity;
        }

        /// <summary>
        /// Change the internal entity and rebind all the attributes.
        /// It's assumed that the attributes of the given entity are exactly
        /// the same as the ones defined.
        /// </summary>
        internal void Rebind(FalkenInternal.falken.EntityBase entity)
        {
            base.Rebind(entity);
            _entity = entity;
        }

        /// <summary>
        /// Load the attributes that are in the internal entity.
        /// </summary>
        internal void LoadEntity(FalkenInternal.falken.EntityBase entity)
        {
            base.LoadAttributes(entity,
              new HashSet<string>() { "position", "rotation" });
            _entity = entity;
        }
    }
}

File: sdk/cpp/src/core/entity.cc
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "falken/entity.h"

#include <string.h>

#include <algorithm>
#include <string>

#include "src/core/attribute_internal.h"
#include "src/core/entity_internal.h"
#include "src/core/globals.h"
#include "src/core/log.h"
#include "absl/memory/memory.h"

namespace falken {

static const AttributeBase::Type kObservationsValidTypes[] = {
    falken::AttributeBase::Type::kTypeFloat,
    falken::AttributeBase::Type::kTypeCategorical,
    falken::AttributeBase::Type::kTypeBool,
    falken::AttributeBase::Type::kTypePosition,
    falken::AttributeBase::Type::kTypeRotation,
    falken::AttributeBase::Type::kTypeFeelers,
    falken::AttributeBase::Type::kTypeUnknown,  // List terminator.
};

namespace {

static const char kEntityTypeName[] = "Entity";

Vector<EntityBase*>::const_iterator FindEntityIteratorByName(
    const Vector<EntityBase*>& entities, const char* name,
    bool log_error) {
  return FindItemIteratorByName(entities, name, log_error,
                                kEntityTypeName);
}

EntityBase* FindEntityByName(const Vector<EntityBase*>& entities,
                             const char* name, bool log_error) {
  return FindItemByName(entities, name, log_error, kEntityTypeName);
}

}  // namespace

// Private data for the EntityBase class.
struct EntityBase::EntityBaseData {
  // Construct entity data with a set of attributes.
  EntityBaseData(EntityContainer* container_, std::string name_)
      : container(container_), name(name_) {}
  // Construct entity data from another one, using a different container.
  EntityBaseData(EntityContainer* container_, const EntityBaseData& other_)
      : container(container_), name(other_.name) {}
  // Construct entity data from another one, using the same container.
  EntityBaseData(EntityBaseData&& other_)
      : container(other_.container), name(std::move(other_.name)) {
    other_.container = nullptr;
  }

  EntityContainer* container;
  std::string name;
};

EntityContainer::EntityContainer() : data_(new EntityContainerData()) {}

EntityContainer::EntityContainer(EntityContainer&& other) = default;

EntityContainer::EntityContainer(
    const std::set<std::string>& entities_to_ignore,
    const EntityContainer& other)
    : data_(new EntityContainerData()) {
  data_->allocated_entities.reserve(other.entities().size());
  for (const EntityBase* entity_base : other.entities()) {
    if (entities_to_ignore.find(std::string(entity_base->name())) ==
        entities_to_ignore.end()) {
      data_->allocated_entities.emplace_back(EntityBase(*this, *entity_base));
    }
  }
}

EntityContainer::~EntityContainer() {
  FALKEN_LOCK_GLOBAL_MUTEX();
  for (EntityBase* entity : data_->entities) {
    entity->ClearContainer();
  }
}

EntityBase::EntityBase(EntityContainer& container, const char* name)
    : AttributeContainer(kObservationsValidTypes, "Observations"),
      FALKEN_UNCONSTRAINED(position),
      FALKEN_UNCONSTRAINED(rotation),
      data_(new EntityBaseData(&container, name)) {
  container.AddEntity(this);
}

EntityBase::EntityBase(EntityContainer& container, const EntityBase& other)
    : AttributeContainer({"position", "rotation"}, other),
      position(*this, other.position),
      rotation(*this, other.rotation),
      data_(new EntityBaseData(&container, *other.data_)) {
  container.AddEntity(this);
}

EntityBase::EntityBase(EntityBase&& other)
    : AttributeContainer({"position", "rotation"}, other),
      position(*this, other.position),
      rotation(*this, other.rotation) {
  other.data_->container->RemoveEntity(&other);
  data_ = absl::make_unique<EntityBaseData>(std::move(*other.data_));
  data_->container->AddEntity(this);
}

EntityBase::~EntityBase() {
  FALKEN_LOCK_GLOBAL_MUTEX();
  if (data_ && data_->container) {
    data_->container->RemoveEntity(this);
  }
}

const char* EntityBase::name() const { return data_->name.c_str(); }

void EntityBase::ClearContainer() { data_->container = nullptr; }

const Vector<EntityBase*>& EntityContainer::entities() const {
  return data_->entities;
}

EntityBase* EntityContainer::entity(const char* name) const {
  return FindEntityByName(data_->entities, name, true);
}

void EntityContainer::RemoveEntity(EntityBase* entity) {
  auto it = FindEntityIteratorByName(data_->entities, entity->name(), false);
  if (it == data_->entities.end()) {
    LogError("Unable to remove entity %s as it is not found in the container.",
             entity->name());
    return;
  }
  data_->entities.erase(it);
}

void EntityContainer::AddEntity(EntityBase* entity) {
  if (FindEntityByName(data_->entities, entity->name(), false)) {
    LogError("Failed to add entity '%s' to the container as an entity with the "
             "same name is already present. Please change the entity name or "
             "use another container.", entity->name());
    return;
  }
  data_->entities.insert(
      std::upper_bound(data_->entities.begin(), data_->entities.end(), entity,
                       [](const EntityBase* lhs, const EntityBase* rhs) {
                         return strcmp(lhs->name(), rhs->name()) < 0;
                       }),
      entity);
}

}  // namespace falken
